go/consensus/cometbft/apps/registry/admission.go

```diff
@@ -0,0 +1,237 @@
+	}
+
+	// Map is present and non-empty, check per-role restrictions
+	// on the maximum number of nodes per entity.
+
+	// Iterate over all valid roles (each entry in the map can
+	// only have a single role).
+	for _, role := range node.Roles() {
+		if !newNode.HasRoles(role) {
+			// Skip unset roles.
+			continue
+		}
+
+		maxNodes, exists := wcfg.MaxNodes[role]
+		if !exists {
+			// No such role found in whitelist.
+			ctx.Logger().Debug("RegisterNode: runtime's whitelist does not allow nodes with given role",
+				"role", role.String(),
+				"runtime_id", rt.ID,
+				"node_id", newNode.ID,
+			)
+			return registry.ErrForbidden
+		}
+		if maxNodes == 0 {
+			// No nodes of this type are allowed.
+			ctx.Logger().Debug("RegisterNode: runtime's whitelist does not allow nodes with given role",
+				"role", role.String(),
+				"runtime_id", rt.ID,
+			)
+			return registry.ErrForbidden
+		}
+
+		if err := verifyNodeCountWithRoleForRuntime(ctx, state, newNode, rt, epoch, role, int(maxNodes)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// verifyNodeCountWithRoleForRuntime verifies that the number of nodes registered by the specified
+// entity for the specified runtime with the specified role is at most the specified maximum.
+func verifyNodeCountWithRoleForRuntime(
+	ctx *api.Context,
+	state *registryState.MutableState,
+	newNode *node.Node,
+	rt *registry.Runtime,
+	epoch beacon.EpochTime,
+	role node.RolesMask,
+	maxNodes int,
+) error {
+	// Count existing nodes owned by entity.
+	nodes, err := state.GetEntityNodes(ctx, newNode.EntityID)
+	if err != nil {
+		ctx.Logger().Error("RegisterNode: failed to query entity nodes",
+			"err", err,
+			"entity_id", newNode.EntityID,
+		)
+		return err
+	}
+
+	var curNodes int
+	for _, n := range nodes {
+		if n.ID.Equal(newNode.ID) || n.IsExpired(uint64(epoch)) || !n.HasRuntime(rt.ID) {
+			// Skip existing node when re-registering.  Also skip
+			// expired nodes and nodes that haven't registered
+			// for the same runtime.
+			continue
+		}
+
+		if n.HasRoles(role) {
+			curNodes++
+		}
+
+		// The check is inside the for loop, so we can stop as
+		// soon as possible once we're over the limit.
+		if curNodes+1 > maxNodes {
+			// Too many nodes with given role already registered.
+			ctx.Logger().Debug("RegisterNode: too many nodes with given role already registered for runtime",
+				"role", role.String(),
+				"runtime_id", rt.ID,
+				"node_id", newNode.ID,
+				"num_registered_nodes", curNodes,
+			)
+			return registry.ErrForbidden
+		}
+	}
+
+	return nil
+}
+
+func verifyRuntimePerRoleAdmissionPolicy(
+	ctx *api.Context,
+	state *registryState.MutableState,
+	newNode *node.Node,
+	rt *registry.Runtime,
+	epoch beacon.EpochTime,
+) error {
+	if len(rt.AdmissionPolicy.PerRole) == 0 {
+		return nil // No per-role policy specified.
+	}
+
+	// Iterate over all valid roles (each entry in the map can only have a single role).
+	for _, role := range node.Roles() {
```

Comment:
I would probably iterate through `rt.AdmissionPolicy.PerRole`, to get rid of the previous `if`.